File: app/src/main/java/pl/laskowski/marcin/model/Region.java
package pl.laskowski.marcin.model;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by Marcin Laskowski.
 */

public class Region implements Iterable<Field> {

    private final Set<Field> fields;

    public Region(Collection<Field> fields) {
        this.fields = new HashSet<>(fields);
    }

    @Override
    public Iterator<Field> iterator() {
        return fields.iterator();
    }

    public boolean isEmpty() {
        return fields.isEmpty();
    }

    public Region intersect(Region other) {
        Region result = new Region(fields);
        result.fields.retainAll(other.fields);
        return result;
    }

    public Region subtract(Region other) {
        Region result = new Region(fields);
        result.fields.removeAll(other.fields);
        return result;
    }

    public boolean isValid() {
        Set<Integer> set = new HashSet<>();
        for (Field field : fields) {
            if (alreadyContainsFieldValue(field, set)) {
                return false;
            } else {
                set.add(field.value());
            }
        }
        return true;
    }

    private boolean alreadyContainsFieldValue(Field field, Set<Integer> numbers) {
        return !field.isEmpty() && numbers.contains(field.value());
    }

    public boolean containsFieldWithSamePosition(Field field) {
        for (Field f : fields) {
            if (f.haveSamePosition(field)) return true;
        }
        return false;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Region region = (Region) o;

        return fields.equals(region.fields);
    }

    @Override
    public int hashCode() {
        return fields.hashCode();
    }

    public int size() {
        return fields.size();
    }

    public List<Field> fullFields() {
        return fields.stream()
                .filter(field -> !field.isEmpty())
                .collect(Collectors.toList());
    }

    public List<Field> emptyFields() {
        return fields.stream()
                .filter(Field::isEmpty)
                .collect(Collectors.toList());
    }

}

File: app/src/main/java/pl/laskowski/marcin/model/Region.kt
package pl.laskowski.marcin.model

data class Region(
    val fields: Set<Field>
): Iterable<Field> {

    constructor(fields: List<Field>): this(fields.toSet())

    override fun iterator() = fields.iterator()

    fun size() = fields.size

    fun isEmpty() = fields.isEmpty()

    fun intersect(other: Region): Region {
        val result = fields.filter { other.fields.contains(it) }
        return Region(result.toSet())
    }

    fun subtract(other: Region): Region {
        val result = fields.filterNot { other.fields.contains(it) }
        return Region(result.toSet())
    }

    fun isValid(): Boolean {
        val set: MutableSet<Int> = HashSet()
        for (field in fields) {
            if (alreadyContainsFieldValue(field, set)) {
                return false
            } else {
                set.add(field.value())
            }
        }
        return true
    }

    fun containsFieldWithSamePosition(field: Field): Boolean {
        return fields.any { it.haveSamePosition(field) }
    }

    fun fullFields(): List<Field> {
        return fields.filterNot { it.isEmpty }
    }

    fun emptyFields(): List<Field> {
        return fields.filter { it.isEmpty }
    }

    private fun alreadyContainsFieldValue(field: Field, numbers: Set<Int>): Boolean {
        return !field.isEmpty && numbers.contains(field.value())
    }
}